qemu/tests/qemu_guest_agent.py

```diff
@@ -923,6 +924,74 @@ def gagent_check_freeze_frozen(self, test, params, env):
             if self.gagent.get_fsfreeze_status() == self.gagent.FSFREEZE_STATUS_FROZEN:
                 self.gagent.fsthaw(check_status=False)
 
+    @error_context.context_aware
+    def gagent_check_hotplug_frozen(self, test, params, env):
+        """
+        hotplug device with frozen fs
+
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment
+        """
+
+        def get_new_disk(disks_before_plug, disks_after_plug):
+            """
+            Get the new added disks by comparing two disk lists.
+            """
+            disk = list(set(disks_after_plug).difference(set(disks_before_plug)))
+            return disk
+
+        session = self._get_session(params, self.vm)
+        image_size_stg0 = params["image_size_stg0"]
+        try:
+            if params.get("os_type") == "linux":
+                disks_before_plug = utils_disk.get_linux_disks(session, True)
+            error_context.context("Freeze guest fs", logging.info)
+            self.gagent.fsfreeze()
+            # For linux guest, waiting for it to be frozen, for windows guest,
+            # waiting for it to be automatically thawed.
+            time.sleep(20)
+            error_context.context("Hotplug a disk to guest", logging.info)
+            image_name_plug = params["images"].split()[1]
+            image_params_plug = params.object_params(image_name_plug)
+            devs = self.vm.devices.images_define_by_params(image_name_plug,
+                                                           image_params_plug,
+                                                           'disk')
+            for dev in devs:
+                self.vm.devices.simple_hotplug(dev, self.vm.monitor)
+            disk_write_cmd = params["disk_write_cmd"]
+            pause = float(params.get("virtio_block_pause", 10.0))
+            error_context.context("Format and write disk", logging.info)
+            if params.get("os_type") == "linux":
+                new_disks = utils_misc.wait_for(lambda: get_new_disk(disks_before_plug.keys(),
+                                                utils_disk.get_linux_disks(session, True).keys()), pause)
+                if not new_disks:
+                    test.fail("Can't get new disks")
+                try:
+                    mnt_point = utils_disk.configure_empty_disk(
+                        session, new_disks[0], image_size_stg0, "linux", labeltype="msdos")
+                except aexpect.ShellTimeoutError:
+                    self.gagent.fsthaw()
+                    mnt_point = utils_disk.configure_empty_disk(
+                        session, new_disks[0], image_size_stg0, "linux", labeltype="msdos")
```

Comment:
@xiangchunfu Please help confirm if this action is what we want(to fsthaw guest if format&mount disk failed and format&mount it again)?

Reply:
@xiangchunfu Get it, thanks.